compiler/rustc_mir_dataflow/src/move_paths/builder.rs

```diff
@@ -390,10 +391,6 @@ impl<'b, 'a, 'tcx> Gatherer<'b, 'a, 'tcx> {
                 self.create_move_path(place);
                 self.gather_init(place.as_ref(), InitKind::Deep);
             }
-
-            TerminatorKind::Drop { place, target: _, unwind: _ } => {
-                self.gather_move(place);
```

Comment:
Would it suffice to just do the first part here (`create_move_path`) instead of having `drop_flag_effects_for_location` to redo this work?

Reply:
Yea, I managed to confuse myself a bit. You're right. It's just a bit scary how far apart the initialization tracking and move checking are.

So the next change would be to change `DropAndReplace` to stop moving out of the dropped value. The only operation we need to do in the move path builder is to gather the operand with which we are replacing the existing value. There's not even a need to handle this case in initialization tracking logic, as it will be initialized right after. Or do you plan to just remove `DropAndReplace` in favor of a `Drop` terminator followed by a regular assignment?

Reply:
> Or do you plan to just remove `DropAndReplace` in favor of a Drop terminator followed by a regular assignment?

yes, that's what I would like to have.
> So the next change would be to change `DropAndReplace` to stop moving out of the dropped value

AFAIK `DropAndReplace` does not move out of the dropped place.
Indeed, in the move path builder `DropAndReplace` already does the exact same thing as an `Assign` (barring the `Operand` / `Rvalue` difference)

Reply:
> DropAndReplace already does the exact same thing as an Assign

Yes, as I understand it, `DropAndReplace` is a hack that was added to work around the following issue: If you treat a drop as being a move, then in this code, `x` is unnecessarily captured by value.

```rust
let mut x = String::new();

|| {
    x = "foo".to_string();
}
```